# Homework 4


### Problem 1 (10% credit). 
What can the principal component decomposition of the regressors help multivariate linear regression with (check all that apply):
a) improve the fit accuracy (R2);
b) remove correlations between regressors;
c) improve the certainty of regression coefficients estimation (shrink confidence intervals for them);
d) allow to better interpret impact of the regressors;
e) increase a number of regressors to gain better prediction power;
f) introduce new regressors that have more impact on the output variable.

In [5]:
import numpy as np
import pandas as pd
import datetime
import matplotlib.pyplot as plt
import scipy
from scipy import stats
%pylab inline
from matplotlib import pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from mpl_toolkits.mplot3d import proj3d
import statsmodels.formula.api as smf

Populating the interactive namespace from numpy and matplotlib


In [6]:
### data for problem 2
data1 = pd.read_csv('https://serv.cusp.nyu.edu/files/ADS-2015/MorningSection/homework4/H41.csv' , index_col=0 )
(N,n)=data1.shape
n=n-1
train1 = data1.loc[0:((N/2)-1),:]
valid1 = data1.loc[(N/2):(N-1),:]
valid1.head()

X1        X2        X3        X4        X5        X6        X7  \
100  0.021073  0.473330 -0.854525 -1.434702 -1.319011  0.602675 -0.627579   
101  0.597511  0.257097 -0.308613 -0.818133  0.362574  0.712588  1.008610   
102  2.152969 -0.739789  0.971372  0.866783 -1.153372 -1.112272  1.221799   
103  0.321259 -0.646392 -0.468620 -0.680378 -0.292803  0.519052  0.093340   
104 -1.134511 -1.250763  0.436301  0.618552 -0.411966  2.915248 -0.735903   

           X8        X9       X10    ...          X12       X13       X14  \
100 -1.546888 -1.434669 -0.003120    ...     0.676649  0.397233 -0.035299   
101  0.210298  0.305236 -0.628875    ...     1.873870  0.032842 -1.207351   
102 -0.325436  0.641371  0.766197    ...    -0.275757  0.561376  0.035587   
103 -0.699125  0.858503  0.081597    ...    -0.798732  0.744467 -0.614448   
104  2.421916 -1.750252 -0.663054    ...    -1.225360 -1.593128 -0.716077   

          X15       X16       X17       X18       X19       X20         Y  
100 -0.426683  1.455772  2.030468 -0.843831 -0.002672 -0.177249 -0.090934  
101 -0.052605  1.215049 -1.272879 -0.684974 -1.042204  1.120010 -2.081685  
102  0.395428 -0.291604 -0.519191 -0.230752  0.903643  1.141344  5.623493  
103  1.687696 -1.866164  1.426627 -0.936641  1.150681 -0.271796 -1.511269  
104 -0.872342 -1.177990 -1.648093 -0.567438 -1.511116  0.013016  3.660144  

[5 rows x 21 columns]

### Problem 2 (40% credit). 
For the data above (training and validation sets) perform a multivariate regression of Y vs all the columns of X and implement a backward stepwise feature selection procedure (evaluating the models trained using different sets of features over the training set using the validation set). Report statistics for both - initial regression and the regression vs the selected subset of regressors.

In [7]:
#data for problems 3-5
data2 = pd.read_csv('https://serv.cusp.nyu.edu/files/ADS-2015/MorningSection/homework4/H42.csv' , index_col=0 )
data2.head()

X1        X2        X3         X4        X5        X6        X7  \
0  10.261135  4.415661  3.185990  14.588875 -2.441283  1.994659 -2.832966   
1  -6.484509  1.496613 -5.139275   8.612433  6.095630  6.515780  2.772823   
2   2.213149  3.328561  2.493297   3.561928  3.432203 -6.106145 -2.403077   
3 -11.539916  1.168906 -5.640197 -12.757397 -3.153445 -5.316729  9.594091   
4   5.227771 -5.799965  1.170432   2.397935  5.572502  0.871541  2.330155   

         X8        X9       X10         Y  
0 -4.060996 -9.552254  2.088611  4.332115  
1 -4.576894 -7.844913 -3.729653 -2.626149  
2  0.254194  2.485210 -4.857989  6.190211  
3  5.387306  3.617259  6.098015 -4.949503  
4 -2.214523  3.428943 -6.646335  2.131978

### Problem 4 (20% of credit)
For the regressors X from the above dataset 2 perform a principal component decomposition and visualize the percent of variation as a function of the number of selected leading principal components.

### Problem 5 (15% of credit)
For the principal component decomposition from the problem 4, visualize the data in a 2d and 3d coordinate system according to the 2 and 3 leading principal components correspondingly.

### Problem 6 (15% of credit)
For the dataset above select a minimal number of leading principal components necessary to cover at least $90\%$ of the entire variation and
perform multivariate regression of the output variable Y vs those selected components.